staging/src/k8s.io/apiserver/pkg/util/flowcontrol/request/mutating_work_estimator.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package request
+
+import (
+	"math"
+	"net/http"
+	"time"
+
+	apirequest "k8s.io/apiserver/pkg/endpoints/request"
+)
+
+func newMutatingWorkEstimator(countFn watchCountGetterFunc) WorkEstimatorFunc {
+	estimator := &mutatingWorkEstimator{
+		countFn: countFn,
+	}
+	return estimator.estimate
+}
+
+type mutatingWorkEstimator struct {
+	countFn watchCountGetterFunc
+}
+
+const (
+	watchesPerSeat                      = 10.0
+	eventAdditionalDurationMilliseconds = 5
+)
+
+func (e *mutatingWorkEstimator) estimate(r *http.Request) WorkEstimate {
+	requestInfo, ok := apirequest.RequestInfoFrom(r.Context())
+	if !ok {
+		// no RequestInfo should never happen, but to be on the safe side
+		// let's return maximumSeats
+		return WorkEstimate{Seats: maximumSeats}
+	}
+
+	watchCount := e.countFn(requestInfo)
+
+	// The cost of the request associated with the watchers of that event
+	// consists of two parts:
+	// - cost of going through the event change logic
+	// - cost of processing an event object for each watcher (e.g. filtering,
+	//     serialization or sending data over network)
+	// We're starting simple to get some operational experience with it and
+	// we will work on tuning the algorithm later. As a starting point we
+	// we simply assume that processing 1 event takes 1/Nth of a seat for
+	// M milliseconds and processing different events is infinitely parallelizable,
+	// so we first try to occupy up to maximumSeats and then adjust additional
+	// latency accordingly.
+	//
+	// TODO: As described in the KEP, we should take into account that not all
+	//   events are equal and try to estimate the cost of a single event based on
+	//   some historical data about size of events.
+	seats := uint(math.Ceil(float64(watchCount) / watchesPerSeat))
+	if seats < minimumSeats {
+		seats = minimumSeats
+	}
+	if seats > maximumSeats {
+		seats = maximumSeats
+	}
+	layers := uint(math.Ceil(float64(watchCount) / (watchesPerSeat * float64(seats))))
```

Comment:
maybe i am getting this wrong? can you clarify on this? if we substitute `seats` with `watchCount/watchesPerSeat` doesn't it always evaluate to 1?

Reply:
This is true only seats was originally < maximumSeats (modulo roundings). If it was higher and we decreased to maximumSeats, this is no longer true.